File: 백준/Gold/9251. LCS/LCS.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        char[] stringA = br.readLine().toCharArray();
        char[] stringB = br.readLine().toCharArray();
        br.close();

        int ALength = stringA.length;
        int BLength = stringB.length;

        int[][] LCS = new int[ALength + 1][BLength + 1];

        LCS[0][0] = 0;
        for(int i = 1; i < ALength + 1; i++) {
            for(int j = 1; j < BLength + 1; j++) {
                if(stringA[i - 1] == stringB[j - 1]) {
                    LCS[i][j] = LCS[i-1][j-1] + 1;
                }
                else {
                    LCS[i][j] = Math.max(LCS[i-1][j], LCS[i][j-1]);
                }
            }
        }
        System.out.println(LCS[ALength][BLength]);
    }
}

File: 백준/Gold/9251. LCS/LCS.py
import sys

input = sys.stdin.readline

X = input().strip() # 처음 문자열
Y = input().strip() # 2번째 문자열

x_length = len(X)
y_length = len(Y)

LCS = [[0] * (y_length+1) for _ in range(x_length+1)] # 최장 공통 부분 수열

for i in range(1, x_length+1):
    for j in range(1, y_length+1):
        
        if X[i-1] == Y[j-1]: # 이전 문자열이 같으면 +1
            LCS[i][j] = LCS[i-1][j-1] + 1
        else: # 아니면 더 큰 LCS 값으로 갱신
            LCS[i][j] = max(LCS[i-1][j], LCS[i][j-1])
            
print(LCS[x_length][y_length])